### PR TITLE
DOE-2897 dashboard repo tagging

### PR DESCRIPTION
**Pull request by apatil-osi**

### CODE REVIEW

**Comment by jeffmaher** [comment]
@apatil-osi : Can you resolve the merge conflicts?

File: Jenkinsfile [hunk 1]
```diff
@@ -1,25 +1,224 @@
...
    catchError {
       stage('Preparation') {
          git branch: 'master', credentialsId: GITHUB_CREDENTIALS_ID, url: GITHUB_URL
       }
       stage('Build'){
          echo 'Nothing to do'
       }
+      stage("Increment Tag") {
+        try {
+            def prEvent = getPullRequestEvent()
```
Comment (jeffmaher) [inline Jenkinsfile]: Can you link to the Jenkins job that will run this? Will be helpful context in order to do a review and if it'll have a pull request event to parse.

File: Jenkinsfile [hunk 2]
```diff
@@ -1,25 +1,224 @@
...
+            newTagPatch++
+            break
+    }
+
+    def newTag = "${newTagMajor}.${newTagMinor}.${newTagPatch}"
+    return newTag
+}
+
+// Updates any build files that contain a version tag
+def updateFiles(String newTag) {
```
Comment (jeffmaher) [inline Jenkinsfile]: Can be removed if not used.

File: Jenkinsfile [hunk 3]
```diff
@@ -1,25 +1,224 @@
...
+            VersionIncrement increment = getVersionIncrement(labels)
+            debug("Increment Tag: increment: ${increment}")
+            if(increment != null ) {
+                def tags = getTags()
+                debug("Increment Tag: tags: ${tags}")
+
+                def newTag = getNewTag(tags, increment)
+                debug("Increment Tag: newTag: ${newTag}")
+
+                updateFiles(newTag)
```
Comment (jeffmaher) [inline Jenkinsfile]: Can be removed if not used.

File: Jenkinsfile [hunk 4]
```diff
@@ -1,25 +1,224 @@
-DOCKER_TAG = '0.6.1-assets'
+/**
+It's hard to test this script without constantly open/closing PRs. 
+Set to true to test with triggering only a build and using mock values.
+*/
+TEST_MODE = false
```
Comment (jeffmaher) [inline Jenkinsfile]: Can remove TEST_MODE related things.

File: Jenkinsfile [hunk 5]
```diff
@@ -1,25 +1,224 @@
...
+
+def debug(String str) {
+    echo "[DEBUG] ${str}"
+}
+
+// Returns Map of Maps containing the parsed JSON from the pull request event
+def getPullRequestEvent() {
+    def prEvents = null
+    
+    if(TEST_MODE) {
```
Comment (jeffmaher) [inline Jenkinsfile]: Can remove TEST_MODE related things.

File: Jenkinsfile [hunk 6]
```diff
@@ -1,25 +1,224 @@
...
 DOCKER_REGISTRY_CREDENTIALS_ID = '6ba8d05c-ca13-4818-8329-15d41a089ec0'
 GITHUB_CREDENTIALS_ID = '433ac100-b3c2-4519-b4d6-207c029a103b'
 GITHUB_URL = 'git@github.com:ca-cwds/dashboard.git'
 GIT_REFSPEC = ''
 
 node('cm-slave') {
+
+    stage('Clone Repo') {
+		  cleanWs()
+		  git branch: 'master', credentialsId: '433ac100-b3c2-4519-b4d6-207c029a103b', url: 'git@github.com:ca-cwds/dashboard.git'
```
Comment (jeffmaher) [inline Jenkinsfile]: Can re-used the GITHUB_URL from above.

File: Jenkinsfile [hunk 7]
```diff
@@ -1,25 +1,224 @@
...
+	}
+
    def image
-   properties([buildDiscarder(logRotator(artifactDaysToKeepStr: '', artifactNumToKeepStr: '', daysToKeepStr: '', numToKeepStr: '3')), disableConcurrentBuilds(), [$class: 'RebuildSettings', autoRebuild: false, rebuildDisabled: false], pipelineTriggers([pollSCM('H/5 * * * *')])])
```
Comment (jeffmaher) [inline Jenkinsfile]: Is this the part that auto-updated the Jenkins settings?

Comment (apatil-osi) [reply Jenkinsfile]: That's correct! Hence, deleted it.

File: Jenkinsfile [hunk 8]
```diff
@@ -1,29 +1,204 @@
...
+        }
+        def pushStatus = sh(script: "${GIT_SSH_COMMAND} git push origin ${newTag}",
+            returnStatus: true)
+        if( pushStatus != 0) {
+            throw new Exception("Unable to push the tag '${newTag}'")
+        }
+    }
+}
+
+//DOCKER_TAG = '0.6.1-assets
```
Comment (jeffmaher) [inline Jenkinsfile]: Is this needed for anything? If not, suggest removing for clarity.

File: Jenkinsfile [hunk 9]
```diff
@@ -1,29 +1,204 @@
...
 DOCKER_NAME = 'cwds/dashboard'
```
Comment (jeffmaher) [inline Jenkinsfile]: Can you move these constants up with the others at the top of the file? A common format to files to keep clean is constants/globals -> methods/functions -> main execution flow

File: Jenkinsfile [hunk 10]
```diff
@@ -1,29 +1,204 @@
...
+
+    return labels
+}
+
+// Takes an array of strings (labels)
+// Returns a VersionIncrement object
+def getVersionIncrement(labels) {
+    debug("getVersionIncrement( labels: ${labels} )")
+
+    def versionIncrement = null
```
Comment (edgibbs) [inline Jenkinsfile]: this can just be `def versionIncrement' it defaults to null.

File: Jenkinsfile [hunk 11]
```diff
@@ -1,29 +1,204 @@
...
+    def versionIncrementsFound = 0
+    for(label in labels){
+        switch(label) {
```
Comment (edgibbs) [inline Jenkinsfile]: The switch here is pretty clear.

File: Jenkinsfile [hunk 12]
```diff
@@ -1,29 +1,204 @@
...
+    }
+
+    def newTag = "${newTagMajor}.${newTagMinor}.${newTagPatch}"
+    return newTag
+}
+
+//Tags the GitHub Repo
+def tagRepo(String newTag) {
+    debug("tagRepo( newTag: ${newTag} )")
+    sshagent (credentials: ['433ac100-b3c2-4519-b4d6-207c029a103b']) {
```
Comment (jeffmaher) [inline Jenkinsfile]: Can you add the credentials to a constant variable? Looks like it's used twice, once here and once in the git clone.

File: Jenkinsfile [hunk 13]
```diff
@@ -1,29 +1,204 @@
...
+}
+
+// Gets all the tags that match SemVer format
+// Returns a list of strings (version number tags)
+def getTags() {
+    def gitTagOutput = sh(script: "git tag", returnStdout: true)
+    debug("getTags(): git tag Output: ${gitTagOutput}")
 
-DOCKER_TAG = '0.6.1-assets'
+    def tags = gitTagOutput.split("\n").findAll{ it =~ /^\d+\.\d+\.\d+$/ }
```
Comment (edgibbs) [inline Jenkinsfile]: can just switch to the implicit return version and drop the tags variable.
```groovy
gitTagOutput.split("\n").findAll{ it =~ /^\d+\.\d+\.\d+$/ }
```

File: Jenkinsfile [hunk 14]
```diff
@@ -1,29 +1,204 @@
...
+	}
+
    def image
-   properties([buildDiscarder(logRotator(artifactDaysToKeepStr: '', artifactNumToKeepStr: '', daysToKeepStr: '', numToKeepStr: '3')), disableConcurrentBuilds(), [$class: 'RebuildSettings', autoRebuild: false, rebuildDisabled: false], pipelineTriggers([pollSCM('H/5 * * * *')])])
    catchError {
       stage('Preparation') {
          git branch: 'master', credentialsId: GITHUB_CREDENTIALS_ID, url: GITHUB_URL
       }
       stage('Build'){
          echo 'Nothing to do'
```
Comment (edgibbs) [inline Jenkinsfile]: Why does this stage exist?

File: Jenkinsfile [hunk 15]
```diff
@@ -1,29 +1,204 @@
...
+            if(increment != null ) {
+                def tags = getTags()
+                debug("Increment Tag: tags: ${tags}")
+
+                def newTag = getNewTag(tags, increment)
+                debug("Increment Tag: newTag: ${newTag}")
+
+                tagRepo(newTag)
+                
+                tagNum = newTag
```
Comment (jeffmaher) [inline Jenkinsfile]: Suggestion: Add `def newTag = null` somewhere outside the `Increment Tag` stage and then no need to create `tagNum`. Benefits: 
1. First creation of variable is easier to spot for maintainability
1. Easier to see that a variable jumps between stages
1. No need for two variables with the same thing (i.e. `tagNum` can be removed)

File: Jenkinsfile [hunk 16]
```diff
@@ -1,29 +1,204 @@
...
+            error "[ERROR] ${ex}"
+        }
+    }
+
       stage ('Build Docker'){
          GIT_REFSPEC = sh(returnStdout: true, script: 'git rev-parse --short HEAD').trim()
          app = docker.build("${DOCKER_NAME}:${GIT_REFSPEC}")
          withDockerRegistry([credentialsId: DOCKER_REGISTRY_CREDENTIALS_ID]) {
-            app.push()
+            app.push(tagNum)
```
Comment (jeffmaher) [inline Jenkinsfile]: See other note about using `newTag`

File: Jenkinsfile [hunk 17]
```diff
@@ -1,29 +1,204 @@
...
 DOCKER_REGISTRY_CREDENTIALS_ID = '6ba8d05c-ca13-4818-8329-15d41a089ec0'
 GITHUB_CREDENTIALS_ID = '433ac100-b3c2-4519-b4d6-207c029a103b'
 GITHUB_URL = 'git@github.com:ca-cwds/dashboard.git'
 GIT_REFSPEC = ''
 
 node('cm-slave') {
+
+    stage('Clone Repo') {
+		  cleanWs()
+		  git branch: 'master', credentialsId: '433ac100-b3c2-4519-b4d6-207c029a103b', url: GITHUB_URL
```
Comment (jeffmaher) [inline Jenkinsfile]: See other note about moving credentials into a constant.

File: Jenkinsfile [hunk 18]
```diff
@@ -1,29 +1,204 @@
...
       stage('Preparation') {
          git branch: 'master', credentialsId: GITHUB_CREDENTIALS_ID, url: GITHUB_URL
       }
       stage('Build'){
          echo 'Nothing to do'
       }
+      stage("Increment Tag") {
+        try {
+            // Link to Jenkins Job: http://jenkins.dev.cwds.io:8080/view/Dashboard/job/dashboard-autotag/
+            def prEvent = getPullRequestEvent()
```
Comment (edgibbs) [inline Jenkinsfile]: I think I'd just name this pullRequestEvent to be explicit.

**Comment by jeffmaher** [comment]
Created an alternative approach that is a little lighter until we can move this into shared library and standardize other projects around this approach. 

See #72 

With that said, it's fine by me to move forward with this PR if you both (@apatil-osi and @edgibbs ) prefer this method and the dev team is aware of how the labels on PRs trigger the builds.

**Comment by edgibbs** [comment]
retest this please

**Pull request merged**